Fix building transmission with C++23

nevack opened the pull request:
Though I didn't check `gtk/` and `qt/`

tearfur reviewed (commented):
Is this exhaustive? I haven't tried, but translatable strings wrapped in `_()` probably needs `fmt::runtime()` too.

nevack commented:
> Is this exhaustive? I haven't tried, but translatable strings wrapped in `_()` probably needs `fmt::runtime()` too.

As I said, I haven't touched UI code in gtk/ and qt/

Only libtransmission and cli are updated

tearfur commented:
There are a bunch of translatable strings in libtransmission too. For example:

https://github.com/transmission/transmission/blob/6c1cee5f79d8bbdc78f96640b52e49441d70247c/libtransmission/announcer-http.cc#L438-L443

nevack commented:
> For example:

My bad, I don't have `HAVE_GETTEXT`, so `#define _(a) (a)`.

We can merge this one as it is, and I will fix other issues in later PRs with the help of my Linux machine.
Or I can convert this one to draft and incrementally fix building any other possible configurations.
I do not have any preference on this.

tearfur commented:
I'd say do it in one go, but you decide whatever works for you :)

mikedld commented:
How are we guaranteeing that this won't break tomorrow, given that we're still on C++17?

nevack commented:
> How are we guaranteeing that this won't break tomorrow, given that we're still on C++17?

I can add workflow for that, but it must be optional for now.

I convert this PR to draft and will add more fixes while thinking how to check this on CI.

nevack commented:
@tearfur You were right, I missed a lot of issues with fmt strings on the first iteration.
Should be fixed now.

tearfur commented on libtransmission/peer-mse.cc:
```diff
@@ -58,8 +58,8 @@ auto export_bits(UIntWide i)
 // NOLINTBEGIN(readability-identifier-naming)
-auto WIDE_INTEGER_CONSTEXPR const generator = wi::key_t{ "2" };
-auto WIDE_INTEGER_CONSTEXPR const prime = wi::key_t{
+auto const generator = wi::key_t{ "2" };
+auto const prime = wi::key_t{
```
Why was this change needed?

nevack replied:
`constexpr`-ness inside this library (wide_integer) fails to compile with C++20.
I think I could try to check upstream for this issue, maybe there's a fix already implemented and a simple dependency update will fix this.

nevack replied:
Actually, works for me now with AppleClang 15.0.0 and LLVM Clang 18.0.6 on macOS.
I think I had problems with GCC 13.x on Linux.

tearfur replied:
Fixed at #7383. You'll want to rebase again after that PR merged, because the GitHub runners now use 24.04 for `ubuntu-latest`, and your last run was on 22.04.

nevack replied:
Thanks, I will rebase after.

For now, this PR is limboed, unfortunately.

nevack replied:
> your last run was on 22.04

Why thou do you think so?

I use `ubuntu-24.04` explicitly in the action introduced in this PR.

tearfur replied:
CodeQL is on `ubuntu-latest`.

Btw do you think it's a good idea to change `ubuntu-24.04` to `ubuntu-latest`?

tearfur replied:
Wait you're right, I wasn't thinking straight. This PR is not blocked by #7383 and feel free to ignore what I said.

I somehow thought the CodeQL CI will fail on the next push because it will run on Ubuntu 24.04 (which has GCC 13 that fails to compile on C++20), but that CI will still be on C++17 so that won't happen.

nevack replied:
Yeah, this was to allow building with `-std=c++20`/`-std=c++23`, not enforcing it on CI.

tearfur commented on qt/Application.h:
```diff
@@ -21,14 +21,14 @@
+#include "Session.h"
+#include "TorrentModel.h"
+#include "WatchDir.h"
 
 class AddData;
-class MainWindow;
-class Prefs;
-class Session;
 class Torrent;
-class TorrentModel;
-class WatchDir;
```
What is this for?

tearfur replied:
Oh now I see the commit message, did compilation fail?

nevack replied:
Yeah, unique_ptr::deleter-s for these fields 
https://github.com/transmission/transmission/blob/main/qt/Application.h#L102-L106
couldn't instantiate because of incomplete types.

mikedld replied:
To be fair, you don't have to replace them with includes. Types must be complete at the time of destruction, meaning you can just add a destructor for `Application` and implement it (even if with `= default`) in .cpp file, where full declarations of those classes are available. Since copy and move constructors/operators are already explicitly deleted, the rule of five will be complete.

nevack replied:
Thanks, Mike, you're right. Done!

tearfur reviewed (changes requested):
Nice that you have made the changes to qt and gtk too!

nevack commented on CMakeLists.txt:
```diff
@@ -170,8 +170,13 @@ set(TR_VCS_REVISION_FILE "${PROJECT_SOURCE_DIR}/REVISION")
 ## Compiler standard version
 
 set(CMAKE_C_STANDARD 11)
```
Done, thanks!

mikedld commented on .github/workflows/actions.yml:
```diff
@@ -990,3 +990,80 @@ jobs:
       run: |
         gradle build
+
+  ubuntu-24-04-from-tarball-cxx-23:
```
Nit: I think we have enough `from-tarball` configurations as is, since this one doesn't exactly need to prove that it builds specifically from a tarball we can speed feedback up by building from git clone instead to not block on tarball generation job.

nevack replied:
Done!

mikedld reviewed (commented):
Overall looks quite intrusive. Since we moved away from `FMT_STRING()` (or whatever it was called) some time ago, which was intrusive as well, I wonder if there's an option in libfmt to differentiate automatically based on the type of argument being either `char const*` or `char const[N]`...

@ckerr, anything specific you had in mind when adding `needs-update` label other than syncing the PR with current `main` to resolve conflicts?

mikedld commented on .github/workflows/actions.yml:
```diff
@@ -988,3 +988,79 @@ jobs:
+            -DRUN_CLANG_TIDY=OFF
+      - name: Make
+        run: cmake --build obj --config RelWithDebInfo -- "-k 0"
+      - name: Test
+        if: ${{ needs.what-to-make.outputs.make-tests == 'true' }}
+        env:
+          TMPDIR: /private/tmp
+        run: cmake -E chdir obj ctest -j $(nproc) --build-config RelWithDebInfo --output-on-failure
+      - name: Install
+        run: cmake --build obj --config RelWithDebInfo --target install/strip
+      - uses: actions/upload-artifact@v4
+        with:
+          name: binaries-${{ github.job }}
+          path: pfx/**/*
```
1. `ENABLE_TESTS` is set to `OFF` above, IIUC tests won't be available
2. Could skip install and artifacts upload too, nothing to check there

I'd remove all 3 steps unless you think we'll be interested in their results. If you agree then (re)building web is also pointless.

nevack replied:
6c330a6591c6a1ca1f355e5f93745c74d58e586c

mikedld commented on CMakeLists.txt:
```diff
@@ -169,9 +169,13 @@ set(TR_VCS_REVISION_FILE "${PROJECT_SOURCE_DIR}/REVISION")
+if(NOT CMAKE_C_STANDARD)
+    set(CMAKE_C_STANDARD 11)
+endif()
 set(CMAKE_C_STANDARD_REQUIRED ON)
```
I'd move `CMAKE_{C,CXX}_STANDARD_REQUIRED` definitions under matching `if(NOT CMAKE_{C,CXX}_STANDARD)` (as we do for e.g. libutp) — if we're not setting the standard, we don't have the right to enforce whatever is set externally and it's better to let it be decided externally as well.

tearfur commented:
Will work in this PR continue? This is a blocker for bumping to C++20.

nevack commented:
> Will work in this PR continue? This is a blocker for bumping to C++20.

Yes, I address raised issues and will update this PR soon

nevack commented on utils/remote.cc:
```diff
@@ -981,7 +981,7 @@ void printDetails(tr_variant* top)
                     if (tr_variantGetStrView(tr_variantListChild(l, child_idx), &sv))
                     {
-                        fmt::print(child_idx == 0 ? "{:s}" : ", {:s}", sv);
+                        fmt::print(fmt::runtime(child_idx == 0 ? "{:s}" : ", {:s}"), sv);
```
6c330a6591c6a1ca1f355e5f93745c74d58e586c

nevack commented:
> > Will work in this PR continue? This is a blocker for bumping to C++20.
> 
> Yes, I address raised issues and will update this PR soon

Rebased, fixed conflicts and applied Mike's suggestions.

Did not test if this is still buildable with c++20/c++23 after months of work being merged.

We have a CI workflow to test this, so I will fix after it will report anything.

nevack commented:
Ready for review

tearfur reviewed (approved):
LGTM generally 👍🏻

nevack commented on .github/workflows/actions.yml:
```diff
@@ -1074,3 +1074,68 @@ jobs:
+            libpsl-dev \
+            libssl-dev \
+            ninja-build \
+            npm
```
030c46ec0d151f8280955904b70ce95e4370b269

nevack commented:
Manually rebased onto latest `main`

ckerr commented:
@mikedld can you re-review? You've still got this PR marked as Changes Requested :)

tearfur commented:
Unfortunately merge conflicts arose again. Hope this is merged soon. 🥲

nevack commented:
@tearfur Could you help me?

I got here build failure after #6914:

<details>
<summary>Details</summary>

```log
FAILED: libtransmission/CMakeFiles/transmission.dir/utils.cc.o 
/usr/bin/clang++ -DFMT_EXCEPTIONS=0 -DFMT_HEADER_ONLY=1 -DHAVE_COPY_FILE_RANGE -DHAVE_FALLOCATE64 -DHAVE_FLOCK -DHAVE_GETMNTENT -DHAVE_GETTEXT -DHAVE_LIBINTL_H -DHAVE_MKDTEMP -DHAVE_NGETTEXT -DHAVE_POSIX_FADVISE -DHAVE_POSIX_FALLOCATE -DHAVE_PREAD -DHAVE_PWRITE -DHAVE_SENDFILE64 -DHAVE_SO_REUSEPORT=1 -DHAVE_STATVFS -DHAVE_SYS_STATVFS_H -DPACKAGE_DATA_DIR=\"/home/runner/work/transmission/transmission/pfx/share\" -DPOSIX -DRAPIDJSON_HAS_STDSTRING=1 -DSMALL_DISABLE_EXCEPTIONS=1 -DWIDE_INTEGER_DISABLE_FLOAT_INTEROP -DWIDE_INTEGER_DISABLE_IOSTREAM -DWIDE_INTEGER_HAS_LIMB_TYPE_UINT64 -DWITH_INOTIFY -DWITH_OPENSSL -DWITH_UTP -D__TRANSMISSION__ -I/home/runner/work/transmission/transmission/src/libtransmission/.. -I/home/runner/work/transmission/transmission/obj/libtransmission/.. -I/home/runner/work/transmission/transmission/src/third-party/libutp/include -I/home/runner/work/transmission/transmission/src/third-party/libb64/include -I/home/runner/work/transmission/transmission/src/third-party/wildmat -isystem /home/runner/work/transmission/transmission/src/third-party/fast_float/include -isystem /home/runner/work/transmission/transmission/obj/third-party/dht.bld/pfx/include -isystem /home/runner/work/transmission/transmission/src/third-party/rapidjson/include -isystem /home/runner/work/transmission/transmission/src/third-party/utfcpp/source -isystem /home/runner/work/transmission/transmission/src/third-party/wide-integer -isystem /home/runner/work/transmission/transmission/src/third-party/fmt/include -isystem /home/runner/work/transmission/transmission/src/third-party/small/include -O2 -g -DNDEBUG -std=gnu++23 -W -Wall -Wextra -Wcast-align -Wextra-semi -Wextra-semi-stmt -Wextra-tokens -Wfloat-equal -Wgnu -Winit-self -Wint-in-bool-context -Wmissing-format-attribute -Wnull-dereference -Wpointer-arith -Wredundant-decls -Wredundant-move -Wreorder-ctor -Wreturn-std-move -Wself-assign -Wself-move -Wsemicolon-before-method-body -Wsentinel -Wshadow -Wsign-compare -Wsometimes-uninitialized -Wstring-conversion -Wsuggest-destructor-override -Wsuggest-override -Wuninitialized -Wunreachable-code -Wunused -Wunused-const-variable -Wunused-parameter -Wunused-result -Wwrite-strings -Wformat-security -MD -MT libtransmission/CMakeFiles/transmission.dir/utils.cc.o -MF libtransmission/CMakeFiles/transmission.dir/utils.cc.o.d -o libtransmission/CMakeFiles/transmission.dir/utils.cc.o -c /home/runner/work/transmission/transmission/src/libtransmission/utils.cc
In file included from /home/runner/work/transmission/transmission/src/libtransmission/utils.cc:10:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/chrono:49:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/shared_ptr.h:53:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/shared_ptr_base.h:59:
/usr/bin/../lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/unique_ptr.h:91:16: error: invalid application of 'sizeof' to an incomplete type '(anonymous namespace)::tr_net_init_impl::tr_net_init_mgr'
   91 |         static_assert(sizeof(_Tp)>0,
      |                       ^~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/unique_ptr.h:398:4: note: in instantiation of member function 'std::default_delete<(anonymous namespace)::tr_net_init_impl::tr_net_init_mgr>::operator()' requested here
  398 |           get_deleter()(std::move(__ptr));
      |           ^
/home/runner/work/transmission/transmission/src/libtransmission/utils.cc:772:52: note: in instantiation of member function 'std::unique_ptr<(anonymous namespace)::tr_net_init_impl::tr_net_init_mgr>::~unique_ptr' requested here
  772 |     static inline std::unique_ptr<tr_net_init_mgr> instance;
      |                                                    ^
/home/runner/work/transmission/transmission/src/libtransmission/utils.cc:740:7: note: definition of '(anonymous namespace)::tr_net_init_impl::tr_net_init_mgr' is not complete until the closing '}'
  740 | class tr_net_init_mgr
      |       ^
1 error generated.
```

</details>

tearfur commented on libtransmission/utils.cc:
```diff
@@ -769,8 +769,11 @@ class tr_net_init_mgr
+    static std::unique_ptr<tr_net_init_mgr> instance;
 };
+
+inline std::unique_ptr<tr_net_init_mgr> tr_net_init_mgr::instance = nullptr;
```
```suggestion
std::unique_ptr<tr_net_init_mgr> tr_net_init_mgr::instance;
```

tearfur reviewed (approved):
Small nit, but otherwise LGTM!

nevack commented:
Test failure on Teamcity for OpenBSD/amd64 seems unrelated.

```
../../../tests/libtransmission/timer-test.cc:60
Expected: (actual) < (upper_bound), actual: 8-byte object <A5-00 00-00 00-00 00-00> vs 8-byte object <96-00 00-00 00-00 00-00>
actual:165 upper_bound:150
../../../tests/libtransmission/timer-test.cc:60
Expected: (actual) < (upper_bound), actual: 8-byte object <A5-00 00-00 00-00 00-00> vs 8-byte object <96-00 00-00 00-00 00-00>
actual:165 upper_bound:150
```

@ckerr Ready to merge, up to you now.

nevack commented:
@ckerr ping!
